egui/src/widgets/plot/items.rs

```diff
@@ -568,3 +662,1032 @@ impl PlotItem for Points {
+            Orientation::Vertical => Value::new(key, value),
+        }
+    }
+
+    fn lower(&self) -> f64 {
+        if self.height.is_sign_positive() {
+            self.base_offset.unwrap_or(0.0)
+        } else {
+            self.base_offset
+                .map(|o| o + self.height)
+                .unwrap_or(self.height)
+        }
+    }
+
+    fn upper(&self) -> f64 {
+        if self.height.is_sign_positive() {
+            self.base_offset
+                .map(|o| o + self.height)
+                .unwrap_or(self.height)
+        } else {
+            self.base_offset.unwrap_or(0.0)
+        }
+    }
+
+    fn bounds_min(&self) -> Value {
+        let key = self.position - self.width / 2.0;
+        let value = self.lower();
+        self.point_at(key, value)
+    }
+
+    fn bounds_max(&self) -> Value {
+        let key = self.position + self.width / 2.0;
+        let value = self.upper();
+        self.point_at(key, value)
+    }
+
+    fn min(&self) -> Value {
+        let key = self.position - self.width / 2.0;
+        let value = self.lower();
+        self.point_at(key, value)
+    }
+
+    fn max(&self) -> Value {
+        let key = self.position + self.width / 2.0;
+        let value = self.upper();
+        self.point_at(key, value)
+    }
+
+    fn base_center(&self) -> Value {
+        self.point_at(self.position, self.base_offset.unwrap_or(0.0))
+    }
+
+    fn value_center(&self) -> Value {
+        self.point_at(
+            self.position,
+            self.base_offset
+                .map(|o| o + self.height)
+                .unwrap_or(self.height),
+        )
+    }
+
+    fn value_right_end(&self) -> Value {
+        self.point_at(
+            self.position + self.width / 2.0,
+            self.base_offset
+                .map(|o| o + self.height)
+                .unwrap_or(self.height),
+        )
+    }
+
+    fn bounds(&self) -> Bounds {
+        let mut bounds = Bounds::NOTHING;
+        bounds.extend_with(&self.bounds_min());
+        bounds.extend_with(&self.bounds_max());
+        bounds
+    }
+
+    fn shapes(&self, transform: &ScreenTransform, highlighted: bool) -> Vec<Shape> {
+        let (stroke, fill) = if highlighted {
+            highlighted_color(self.stroke, self.fill)
+        } else {
+            (self.stroke, self.fill)
+        };
+
+        let mut shapes = Vec::new();
+
+        let rect = transform.rect_from_values(&self.min(), &self.max());
+        let rect = Shape::Rect {
+            rect,
+            corner_radius: 0.0,
+            fill,
+            stroke,
+        };
+        shapes.push(rect);
+        shapes
+    }
+
+    fn default_values_format(&self, transform: &ScreenTransform) -> String {
+        let mut text = String::new();
+        let scale = transform.dvalue_dpos();
+        let y_decimals = ((-scale[1].abs().log10()).ceil().at_least(0.0) as usize).at_most(6);
+        text.push_str(&format!("\n{:.*}", y_decimals, self.height));
```

Comment:
Why not just return the format string here?

Reply:
Just noticed: Same goes for lines 1361-1365. Write just one format string since it may reduce allocations.